java-frontend/src/main/java/org/sonar/java/resolve/Types.java

```diff
@@ -194,4 +246,135 @@ public int compare(Type type, Type t1) {
+    reducedList.addAll(types.subList(2, types.size()));
+
+    return leastContainingParameterization(reducedList);
+  }
+
+  private Type leastContainingTypeArgument(JavaType type1, JavaType type2) {
+    Preconditions.checkArgument(type1.isTagged(JavaType.PARAMETERIZED) && type2.isTagged(JavaType.PARAMETERIZED));
+
+    TypeSubstitution typeSubstitution1 = ((ParametrizedTypeJavaType) type1).typeSubstitution;
+    TypeSubstitution typeSubstitution2 = ((ParametrizedTypeJavaType) type2).typeSubstitution;
+
+    TypeSubstitution newTypeSubstitution = new TypeSubstitution();
+    for (TypeVariableJavaType typeVar : typeSubstitution1.typeVariables()) {
+      JavaType subs1 = typeSubstitution1.substitutedType(typeVar);
+      JavaType subs2 = typeSubstitution2.substitutedType(typeVar);
+
+      boolean isWildcard1 = subs1.isTagged(JavaType.WILDCARD);
+      boolean isWildcard2 = subs2.isTagged(JavaType.WILDCARD);
+
+      JavaType newSubs;
+      if (subs1 == subs2) {
+        newSubs = subs1;
+      } else if (isWildcard1 && isWildcard2) {
+        newSubs = lctaBothWildcards((WildCardType) subs1, (WildCardType) subs2);
+      } else if (isWildcard1 ^ isWildcard2) {
+        JavaType rawType = isWildcard1 ? subs2 : subs1;
+        WildCardType wildcardType = (WildCardType) (isWildcard1 ? subs1 : subs2);
+        newSubs = lctaOneWildcard(rawType, wildcardType);
+      } else {
+        // subs1 and subs2 are not wildcards
+        JavaType newBound = (JavaType) cachedLeastUpperBound(Sets.newHashSet(subs1, subs2));
+        newSubs = parametrizedTypeCache.getWildcardType(newBound, BoundType.EXTENDS);
+      }
+      newTypeSubstitution.add(typeVar, newSubs);
+    }
+
+    return parametrizedTypeCache.getParametrizedTypeType(type1.symbol, newTypeSubstitution);
+  }
+
+  private JavaType lctaOneWildcard(JavaType rawType, WildCardType wildcardType) {
+    if (wildcardType.boundType == BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Sets.newHashSet(rawType, wildcardType.bound));
+      return parametrizedTypeCache.getWildcardType(glb, BoundType.SUPER);
+    }
+    JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(rawType, wildcardType.bound));
+    return parametrizedTypeCache.getWildcardType(lub, BoundType.EXTENDS);
+  }
+
+  private JavaType lctaBothWildcards(WildCardType type1, WildCardType type2) {
+    if (type1.boundType == BoundType.SUPER && type2.boundType == BoundType.SUPER) {
+      JavaType glb = (JavaType) greatestLowerBound(Sets.newHashSet(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(glb, BoundType.SUPER);
+    }
+    if (type1.boundType == BoundType.EXTENDS && type2.boundType == BoundType.EXTENDS) {
+      JavaType lub = (JavaType) cachedLeastUpperBound(Sets.newHashSet(type1.bound, type2.bound));
+      return parametrizedTypeCache.getWildcardType(lub, BoundType.EXTENDS);
+    }
+    if (type1.bound == type2.bound) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)